src/openads/domain/service/DomainEventBus.js

```diff
@@ -0,0 +1,45 @@
+class DomainEventBus {
+  constructor () {
+    this._observers = new Map()
+  }
+
+  register ({eventName, observer}) {
+    if (!eventName) {
+      throw new Error('Event Name is required')
+    }
+    if (typeof observer !== 'function') {
+      throw new Error('Observer must be a function')
+    }
+    if (!this._observers.has(eventName)) {
+      this._observers.set(eventName, [observer])
+    } else {
+      this._observers.get(eventName).push(observer)
+    }
+  }
+
+  raise ({domainEvent}) {
+    this._observers
+      .get(domainEvent.eventName)
+      .forEach(observer => {
+        try {
+          observer({
+            payload: domainEvent.payload,
+            dispatcher: (data) => this.raise({domainEvent: data})
+          })
+        } catch (err) {
+          console.log('Error processing the observer: ', err)
+        }
+      })
+  }
+
+  getObservers () {
```

Comment:
giving access to the observers could break them out of the DomainEventBus

Reply:
totally agree, it's private state, we should only expose for public use, raise, register and clear

Reply:
Well spotted! Right now the only meaning of this method is for test purposes so I can expose two new 'informative' methods with the number of registered events and the number of observers registered for a particular event. With this scenario, I will be able to retrieve the needed info for testing/check purposes and avoid exposing the internal map.

Reply:
Done
![](https://media.giphy.com/media/3oKIPf3C7HqqYBVcCk/giphy.gif)